### PR TITLE
allow Sink.queue concurrent pulling

### PR DESCRIPTION
**Pull request by yakivy**
## Purpose

Allow to use Sink.queue with multiple consumers.

## Changes

The main change is that `currentRequest` currently holds the queue with input requests. Size limit is also taken into account.
```
var currentRequests: mutable.Queue[Requested[T]] = mutable.Queue.empty
```

## Background Context

I faced that `Sink.queue` doesn't work with multiple consumers and it looks like I found the easy approach how to make it possible. Just decided to share my local fix.

### Ref: #27361

### CODE REVIEW

**Comment by raboof** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 1]
```diff
@@ -351,7 +356,7 @@ import org.reactivestreams.Subscriber
       require(maxBuffer > 0, "Buffer size must be greater than 0")
 
       var buffer: Buffer[Received[T]] = _
-      var currentRequest: Option[Requested[T]] = None
+      var currentRequests: mutable.Queue[Requested[T]] = mutable.Queue.empty
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Use `Buffer` here as well, with the max size

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 2]
```diff
@@ -363,17 +368,13 @@ import org.reactivestreams.Subscriber
...
-                  "You have to wait for previous future to be resolved to send another request"))
-            case None =>
-              if (buffer.isEmpty) currentRequest = Some(pullPromise)
-              else {
-                if (buffer.used == maxBuffer) tryPull(in)
-                sendDownstream(pullPromise)
-              }
+          if (currentRequests.size >= requestBufferSize)
+            pullPromise.failure(new IllegalStateException(
+              "Request buffer is full. You have to wait for one previous future to be resolved to send another request"))
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: That there is a request buffer is an implementation concern, what is really happening is that there are too many concurrent requests. I think the error message could also include the buffer size as the max number of concurrent `SinkQueue.pull`s.

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 3]
```diff
@@ -308,6 +308,26 @@ object Sink {
...
+   * `Sink` will request at most number of elements equal to size of `inputBuffer` from
+   * upstream and then stop back pressure.  You can configure size of input
+   * buffer by using [[Sink.withAttributes]] method.
+   *
+   * For stream completion you need to pull all elements from [[akka.stream.javadsl.SinkQueueWithCancel]] including last None
+   * as completion marker
+   *
+   * @see [[akka.stream.javadsl.SinkQueueWithCancel]]
+   */
+  def queue[T](requestBufferSize: Int): Sink[T, SinkQueueWithCancel[T]] =
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Maybe call it `maxConcurrentPulls` or perhaps `parallelism` like `mapAsync` rather, that it is a buffer size is an internal concern.

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 4]
```diff
@@ -308,6 +308,26 @@ object Sink {
...
+   * Creates a `Sink` that is materialized as an [[akka.stream.javadsl.SinkQueueWithCancel]].
+   * [[akka.stream.javadsl.SinkQueueWithCancel.pull]] method is pulling element from the stream and returns ``CompletionStage[Option[T]]``.
+   * `CompletionStage` completes when element is available.
+   *
+   * Before calling pull method second time need to ensure that request buffer is not full or wait until some of the previous CompletionStages completes.
+   * Pull returns Failed future with ''IllegalStateException'' if request buffer is full.
+   *
+   * `Sink` will request at most number of elements equal to size of `inputBuffer` from
+   * upstream and then stop back pressure.  You can configure size of input
+   * buffer by using [[Sink.withAttributes]] method.
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Hmm I wonder why it should have a buffer at all, you can always compose with a buffer upstream. Would simplify the implementation a lot.

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 1 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Fixed

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 2 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Fixed

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 3 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Fixed

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 4 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Totally agree, left only for backward compatibility.

**Comment by yakivy** [comment]
@johanandren I also faced with the same feature requirement for `Source.queue`, so added concurrency support there.

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 4 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: I've created #28270 to track that, no need to change here.

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala (hunk 3 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: Fixed in the Scaladsl but still `requestBufferSize` here.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 5]
```diff
@@ -812,7 +812,44 @@ object Source {
...
+   * @param overflowStrategy Strategy that is used when incoming elements cannot fit inside the buffer
+   * @param maxConcurrentOffers maximum number of pending offers for backpressure overflow strategy
+   *                            when buffer is full
+   */
+  def queue[T](
+      bufferSize: Int,
+      overflowStrategy: OverflowStrategy,
+      maxConcurrentOffers: Int): Source[T, SourceQueueWithComplete[T]] =
+    Source.fromGraph(
+      new QueueSource(bufferSize, overflowStrategy, maxConcurrentOffers).withAttributes(DefaultAttributes.queueSource))
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: Please move the Source queue change to a separate PR to simplify reviewing.

Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: Excuse me, but how critical is this? I mixed changes for both `Source.queue` and `Sink.queue` in my last 3 commits, so decoupling can me done only by manual changes.

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: I think Source will need more careful consideration, while the sink is basically done for merge. It's completely different files so shouldn't be hard to pick those out of a diff?

Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: Got it. Fixed and created:
https://github.com/akka/akka/issues/28272
https://github.com/akka/akka/pull/28273

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: Thanks

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 6]
```diff
@@ -311,8 +311,13 @@ import scala.util.control.NonFatal
 /**
  * INTERNAL API
  */
-@InternalApi private[akka] final class QueueSink[T]()
+@InternalApi private[akka] final class QueueSink[T](maxConcurrentPulls: Int)
     extends GraphStageWithMaterializedValue[SinkShape[T], SinkQueueWithCancel[T]] {
+
+  def this() = this(1)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: I think this isn't used, so remove it. The delegation in the Flow dsl is a better idea anyway.

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 7]
```diff
@@ -329,29 +334,27 @@ import scala.util.control.NonFatal
...
 
       var buffer: Buffer[Received[T]] = _
-      var currentRequest: Option[Requested[T]] = None
+      var currentRequests: Buffer[Requested[T]] = _
 
       override def preStart(): Unit = {
         // Allocates one additional element to hold stream
         // closed/failure indicators
         buffer = Buffer(maxBuffer + 1, inheritedAttributes)
+        currentRequests = Buffer(maxConcurrentPulls, inheritedAttributes)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: I think this is a left over of some old problem with buffers, please move both buffer creations to the constructor and make the fields `val` instead of var.

**Review by johanandren — COMMENTED** [review]
I think the logic looks alright but there is one more thing to cover - making sure to complete or fail all waiting promises when the stage completes or fails, and also on abrupt termination (`Materializer` or `ActorSystem` terminated).

The first two you can hook into sendDownstream when dequeueing `Some(None)` or `Failure`. The abrupt you have to do in a `postStop`, neither of the `onUpstream` callbacks are triggered in that case, since `postStop` will always be executed the complete/fail ones should dequeue promises from `currentRequests` and then `postStop` can just loop through any promises left in the buffer and fail them with a `akka.stream.AbruptStageTerminationException`

Would be good to cover these cases with tests also since it seems to be missing.
For an example of testing the abrupt termination case, you can look at `akka.stream.scaladsl.SeqSinkSpec`

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 6 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Fixed

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 7 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Fixed

**Comment by yakivy** [comment]
OK TO TEST

**Review by johanandren — APPROVED** [review]
LGTM with one tiny fix

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 8]
```diff
@@ -311,8 +311,11 @@ import scala.util.control.NonFatal
 /**
  * INTERNAL API
  */
-@InternalApi private[akka] final class QueueSink[T]()
+@InternalApi private[akka] final class QueueSink[T](maxConcurrentPulls: Int)
     extends GraphStageWithMaterializedValue[SinkShape[T], SinkQueueWithCancel[T]] {
+
+  require(maxConcurrentPulls > 0, "Request buffer size must be greater than 0")
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: wrong error message?

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 9]
```diff
@@ -366,23 +364,28 @@ import scala.util.control.NonFatal
...
         if (buffer.used < maxBuffer) pull(in)
       }
 
-      override def onUpstreamFinish(): Unit = enqueueAndNotify(Success(None))
-      override def onUpstreamFailure(ex: Throwable): Unit = enqueueAndNotify(Failure(ex))
+      override def onUpstreamFinish(): Unit = {
+        buffer.enqueue(Success(None))
+        while (currentRequests.nonEmpty && buffer.nonEmpty) currentRequests.dequeue().complete(buffer.dequeue())
+        while (currentRequests.nonEmpty) currentRequests.dequeue().complete(Success(None))
+        if (buffer.isEmpty) completeStage()
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: when/how will the stage be completed if this `if` is false?

**Review by patriknw — COMMENTED** [review]
seems good to add this capability

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 9 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: When subscribers register they will immediately consume what's in the buffer, if it reaches the Success(None) of Failure(ex) it will complete/fail the stage.

I wonder though for failure if it shouldn't just fail immediately.

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 10]
```diff
@@ -328,30 +331,25 @@ import scala.util.control.NonFatal
       val maxBuffer = inheritedAttributes.get[InputBuffer](InputBuffer(16, 16)).max
       require(maxBuffer > 0, "Buffer size must be greater than 0")
 
-      var buffer: Buffer[Received[T]] = _
-      var currentRequest: Option[Requested[T]] = None
+      // Allocates one additional element to hold stream closed/failure indicators
+      var buffer: Buffer[Received[T]] = Buffer(maxBuffer + 1, inheritedAttributes)
+      var currentRequests: Buffer[Requested[T]] = Buffer(maxConcurrentPulls, inheritedAttributes)
```
[the remark this answers is not in the document]
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Fixed

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 8 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: Yes, fixed.

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 9 above)
Comment (yakivy) [reply akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: These are the handles for the upstream, elements that are already passed to the queue should not depend from upstream and we allow the queue to keep going until there are elements in the buffer.

**Comment by yakivy** [comment]
@johanandren travis started checking backward compatibility for `@InternalApi private[akka]` classes, is it expected?

**Comment by johanandren** [comment]
@yakivy I'm afraid jenkins/mima doesn't understand the annotation so you'll have to add a mima exclude for the change.

**Comment by yakivy** [comment]
hm... but everything was ok on previous builds. Ok, I'll add exclude.

**Review by johanandren — APPROVED** [review]

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
PLS BUILD

**Comment by patriknw** [comment]
can be merged if additional build is successful
